### DISCUSSION TITLE
SDFS "OTA" Updates

### DISCUSSION THREAD

**Discussion by NuclearPhoenixx** [earlephilhower/arduino-pico#1961] [Q&A] [answered]
Hi there, the title is pretty self-explanatory: is it possible to use the [Updater](https://github.com/earlephilhower/arduino-pico/blob/master/libraries/Updater/src/Updater.h) library (Updater.h) to pull firmware files from SDFS and apply them like with a "normal" OTA update, just skipping the over-the-air part? ;)

Given you can pull a `streamVar` from the SD card is it possibly just as simple as using

```cpp
Update.begin(firmwareLengthInBytes);
Update.writeStream(streamVar);
Update.end();
```

like described in the docs or am I completely on the wrong track here?

Please let me know if something similar was already asked here, I didn't find anything related on my search.

**Comment by earlephilhower** [comment, marked as answer]
Yes, it is possible to read a file from SDFS, send it to the Updater who will then write it to onboard flash as a file in LittleFS, and then reboot and write from onboard LittleFS to onboard flash.

If you mean, can you directly use SDFS files to update w/o an onboard LittleFS partition, then that is not possible to do today.  The actual OTA is handled [here](https://github.com/earlephilhower/arduino-pico/tree/master/ota), and it doesn't know anything about SPI or the FAT filesystem.

To make that work, you'd need to make a minimal bare-metal (non-Arduino) FAT filesystem driver, a minimal (non-Arduino) SPI driver, and define a new partition table structure.  It's definitely possible, but not a quick job I would think

**Reply by NuclearPhoenixx** [reply]
> Yes, it is possible to read a file from SDFS, send it to the Updater who will then write it to onboard flash as a file in LittleFS, and then reboot and write from onboard LittleFS to onboard flash.

Perfect, that's exactly what I want to do. I have a 1MB sketch and 3MB LittleFS partition and want users to be able to update by plugging in an SD card, eliminating the need for an active computer connection. Thanks a bunch!
